### ISSUE TITLE
NOT_ALLOWED error page allows embedding of unquoted content from URL and method

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4978]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#4978|
|Type|enhancement|
|Created|2011-03-22 20:22:26Z|
|Branch|https://github.com/twisted/twisted/tree/not-implemented-quote-4978|

All browsers quote this stuff except, surprise, IE6.  This is therefore vulnerable on that browser to XSS attacks.

<details><summary>Searchable metadata</summary>

```
trac-id__4978 4978
type__enhancement enhancement
reporter__glyph glyph
priority__high high
milestone__ 
branch__branches_not_implemented_quote_4978 branches/not-implemented-quote-4978
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__web web
keywords__security security
time__1300825346000000 1300825346000000
changetime__1300830361000000 1300830361000000
version__None None
owner__glyph glyph
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

(In [31385]) Branching to 'not-implemented-quote-4978'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph removed owner|
|-|-|

[Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/not-implemented-quote-4978)
[Diff](http://twistedmatrix.com:12435/4978)

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_4"></a>|@mwhudson set owner to @glyph|
|-|-|

I don't quite understand the selective vertical whitespace fixes in test_web, aren't all test methods supposed to be separated by two blank lines?

Also, I hate assert{Not,}In(thing, <blob of html>) tests, but I guess that's appropriate here.

All said, looks good, please land.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

Replying to [mwh](#note_4):
> I don't quite understand the selective vertical whitespace fixes in test_web, aren't all test methods supposed to be separated by two blank lines?

As far as I can tell, the only whitespace fixes I made did exactly that.  One line had some trailing whitespace, and I stripped it.  Maybe you were misreading that as adding an additional (wrong) blank line?

> Also, I hate assert{Not,}In(thing, <blob of html>) tests, but I guess that's appropriate here.

Yeah, when I re-do this functionality with `twisted.web.template`, it will be much better.

> All said, looks good, please land.

Thank you.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph set status to `closed`|
|-|-|

(In [31389]) Merge not-implemented-quote-4978

Author: glyph

Reviewer: mwh

Fixes: [#4978](https://github.com/twisted/twisted/issues/4978)

default NOT_IMPLEMENTED and NOT_ALLOWED pages now quote the request method and
URI respectively, to protect against browsers which don't quote those values for
us.
